Fix floating point arithmetic issues in EqualWidthHistogram

timzimmermann opened the pull request:
I encountered another nasty rounding error for floating point data types in histograms.
This fix introduces an early exit to avoid that error.

Bouncner commented:
Could you elaborate or refer to a line in the code where the previous error occurred?
You test for no_throw. I'd have expected that the histogram never throws for such requests?!?

timzimmermann commented:
> Could you elaborate or refer to a line in the code where the previous error occurred?
> You test for no_throw. I'd have expected that the histogram never throws for such requests?!?

I expected that, too, but the function I fixed previously could have returned an invalid BinID.
If you remove the if clause I added, it will throw, because the returned BinID is out of bounds, and the `DebugAssert` in `bin_height()` will throw.
We could test for the specific value, but the value doesn't really matter.

The problem is in those lines further down in the function:

```c++
if (_bin_data.bin_count_with_larger_range == 0u ||
         value <= _bin_maximum(_bin_data.bin_count_with_larger_range - 1u)) {
  // All bins up to that point have the exact same width, so we can use index 0.
  return (value - _bin_data.minimum) / _bin_width(0u);
}
```

That's the path that will always be chosen for floats, because `_bin_data.bin_count_with_larger_range == 0` for all floating point histograms.
Now, because of floating point arithmetic, the calculation for the return value could calculate a BinID that is out of bounds.
Specifically, if we have 10,000 bins, the function might return 10,000.
That's exactly what would happen in the test I added.

timzimmermann commented:
I added another fix + test for a similar issue, and added some comments in the relevant places for clarification.
Does that make more sense, @Bouncner?

mrks commented on src/lib/statistics/chunk_statistics/histograms/equal_width_histogram.cpp:
```diff
@@ -270,7 +270,13 @@ BinID EqualWidthHistogram<T>::_bin_for_value(const T& value) const {
+      // The above calculation can lead to an index that is equal to or larger than the number of bins there are,
+      // due to floating point arithmetic.
+      // We checked before that the value is not larger than the maximum of the histogram,
+      // so simply return the last bin.
```
"... so in that case, simply return the last bin"?

timzimmermann replied:
Done.

Bouncner commented on src/test/statistics/chunk_statistics/histograms/equal_width_histogram_test.cpp:
```diff
@@ -507,6 +507,24 @@ TEST_F(EqualWidthHistogramTest, FloatLessThan) {
+}
+
+TEST_F(EqualWidthHistogramTest, FloatBinBoundariesLargeValues) {
+  const auto value = 501506.55f;
```
'-notation?

timzimmermann replied:
Yep.

Bouncner commented on src/test/statistics/chunk_statistics/histograms/equal_width_histogram_test.cpp:
```diff
@@ -507,6 +507,24 @@ TEST_F(EqualWidthHistogramTest, FloatLessThan) {
+  const auto bin_count = 10000u;
+  const auto hist = EqualWidthHistogram<float>(min, max, std::vector<HistogramCountType>(bin_count, 1u),
+                                               std::vector<HistogramCountType>(bin_count, 1u), 0u);
+  EXPECT_NO_THROW(hist.estimate_cardinality(PredicateCondition::GreaterThanEquals, max));
```
That's not what I meant. If we think we should test for NOTHROW, there needs to be a very good explanation, because this test could be done for every single function in Hyrise.
I would see the previous behaviour as a bug that is hard to find out. If we want to have a test for that bug, there should be a comment explaining that.

timzimmermann replied:
Sorry for the delay, please check whether the comments make sense to you.

Bouncner commented on src/test/statistics/chunk_statistics/histograms/equal_width_histogram_test.cpp:
```diff
@@ -518,6 +521,15 @@ TEST_F(EqualWidthHistogramTest, FloatRounding) {
+  // calculate the bin boundaries, while we divided by the bin_width to calculate the
+  // boundaries for a given bin in _bin_for_value().
+  // Adding the bin_width (a float for float histograms) in a loop introduces an error
+  // that increses in every iteration due to floating point arithmetic.
```
"Increases".

'-notation for numbers.
